apex/contrib/csrc/group_norm/group_norm_nhwc_op.cpp

```diff
@@ -0,0 +1,243 @@
+/***************************************************************************************************
+ * Copyright (c) 2011-2023, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ *modification, are not permit- ted.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY DIRECT,
+ *INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ *DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ *OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ *NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+ *EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ **************************************************************************************************/
+
+#include <ATen/cuda/CUDAContext.h>
+#include <torch/extension.h>
+
+#include "group_norm_nhwc_fwd_one_pass.h"
+#include "group_norm_nhwc_bwd_one_pass.h"
+#include "group_norm_nhwc.h"
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#define CHECK_CUDA_STATUS(call)                                       \
+  do {                                                                \
+    cudaError_t status_ = call;                                       \
+    if (status_ != cudaSuccess) {                                     \
+      fprintf(stderr, "CUDA error (%s:%d): %s\n", __FILE__, __LINE__, \
+              cudaGetErrorString(status_));                           \
+      exit(1);                                                        \
+    }                                                                 \
+  } while (0)
+
+#define CHECK_CUDA(x) \
+  AT_ASSERTM(x.type().is_cuda(), #x " must be a CUDA tensor")
+#define CHECK_CONTIGUOUS(x) \
+  AT_ASSERTM(x.is_contiguous(), #x " must be contiguous")
+#define CHECK_CHANNELS_LAST(x)                                \
+  AT_ASSERTM(x.is_contiguous(at::MemoryFormat::ChannelsLast), \
+             #x " must be channels last")
+#define CHECK_INPUT(x) \
+  CHECK_CUDA(x);       \
+  CHECK_CONTIGUOUS(x)
+#define CHECK_NHWC_INPUT(x) \
+  CHECK_CUDA(x);            \
+  CHECK_CHANNELS_LAST(x)
+
+enum Algo { OnePass, TwoPass };
+
+static bool initialized = false;
+static cudaDeviceProp props;
```

Comment:
I prefer to spell out the supported `C` and `groups` and check the arguments of fwd and bwd like
```c++
static cudaDeviceProp props;
constexpr std::unordered_set<int> supported_c_values = {128, 256, 320, 448, 512, 640, 768, 896, 960, 1024, 1280, 1344, 1536, 1792, 1920, 2048, 2240, 2560, 2688, 3072, 3136, 3584, 4096};
constexpr std::unordered_set<int> supported_groups_values = {32};
```

Reply:
Ditto.